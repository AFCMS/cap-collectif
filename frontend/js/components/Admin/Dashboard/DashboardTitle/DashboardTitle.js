// @flow
import * as React from 'react';
import { graphql, useFragment } from 'react-relay';
import { useIntl } from 'react-intl';
import Flex from '~ui/Primitives/Layout/Flex';
import type { DashboardTitle_query$key } from '~relay/DashboardTitle_query.graphql';
import Text from '~ui/Primitives/Text';
import { headingStyles } from '~ui/Primitives/Heading';
import { FontWeight } from '~ui/Primitives/constants';

type Props = {|
  query: DashboardTitle_query$key,
|};

const FRAGMENT = graphql`
  fragment DashboardTitle_query on Query {
    allProject: projects {
      totalCount
    }
    inProgressProjects: projects(status: 1) {
      totalCount
    }
    doneProjects: projects(status: 2) {
      totalCount
    }
  }
`;

const DashboardTitle = ({ query: queryFragment }: Props): React.Node => {
  const query = useFragment(FRAGMENT, queryFragment);
  const intl = useIntl();
  const { allProject, inProgressProjects, doneProjects } = query;

  return (
    <Flex direction="row" align="center" justify="space-between" px={6} py={4} bg="white">
      <Text color="blue.800" {...headingStyles.h4} fontWeight={FontWeight.Semibold}>
        {intl.formatMessage({ id: 'dashboard-platform' })}
      </Text>

      <Flex
        direction="row"
        color="blue.800"
        spacing={4}
        {...headingStyles.h5}
        fontWeight={FontWeight.Semibold}>
        <Text>
          <Text as="span" color="blue.500" mr={1}>
            {allProject.totalCount}
          </Text>
          {intl.formatMessage({ id: 'global.project-dynamic' }, { num: allProject.totalCount })}
        </Text>

        <Text>
          <Text as="span" color="orange.500" mr={1}>
            {inProgressProjects.totalCount}
          </Text>
          {intl.formatMessage({ id: 'global.in-progress' })}
        </Text>

        <Text>
          <Text as="span" color="green.500" mr={1}>
            {doneProjects.totalCount}
          </Text>

          {intl.formatMessage({ id: 'global.done-dynamic' }, { num: doneProjects.totalCount })}
        </Text>
      </Flex>
    </Flex>
  );
};

export default DashboardTitle;
