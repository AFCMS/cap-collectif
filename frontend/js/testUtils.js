// @flow
import * as React from 'react';
import ReactDOM from 'react-dom';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl-redux';
import { RelayEnvironmentProvider } from 'react-relay';
import { theme } from '~/styles/theme';
import appStore from '~/stores/AppStore';

export const mockRandomValues = () => {
  global.Math.random = () => 0.5;
};

export const addsSupportForPortals = () => {
  // See: https://github.com/facebook/react/issues/11565
  // $FlowFixMe
  ReactDOM.createPortal = jest.fn(element => {
    return element;
  });
};

export const clearSupportForPortals = () => {
  // $FlowFixMe
  ReactDOM.createPortal.mockClear();
};

type Props = {| +children: React.Node, +store?: any |};

export const MockProviders = ({ children, store = {} }: Props) => {
  return (
    <Provider store={appStore(store)}>
      <IntlProvider>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </IntlProvider>
    </Provider>
  );
};

type RelaySuspensFragmentTestProps = {| environment: any, ...Props |};

export const RelaySuspensFragmentTest = ({
  children,
  environment,
  store = {},
}: RelaySuspensFragmentTestProps) => {
  return (
    <MockProviders store={store}>
      <RelayEnvironmentProvider environment={environment}>
        <React.Suspense>{children}</React.Suspense>
      </RelayEnvironmentProvider>
    </MockProviders>
  );
};

export default MockProviders;
