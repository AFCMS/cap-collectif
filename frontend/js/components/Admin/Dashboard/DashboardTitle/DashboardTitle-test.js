// @flow
/* eslint-env jest */
import * as React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { graphql, useLazyLoadQuery } from 'react-relay';
import { createMockEnvironment, MockPayloadGenerator } from 'relay-test-utils';
import { RelaySuspensFragmentTest } from '~/testUtils';
import type { DashboardTitleTestQuery } from '~relay/DashboardTitleTestQuery.graphql';
import DashboardTitle from './DashboardTitle';

describe('<DashboardTitle />', () => {
  let environment;
  let testComponentTree;
  let TestDashboardTitle;

  const query = graphql`
    query DashboardTitleTestQuery @relay_test_operation {
        viewer {
            ...DashboardTitle_viewer
        }
    }
  `;

  const defaultMockResolvers = {
    User: () => ({
      allProject: {
        totalCount: 10,
      },
      inProgressProjects: {
        totalCount: 5,
      },
      doneProjects: {
        totalCount: 5,
      },
    }),
  };

  beforeEach(() => {
    environment = createMockEnvironment();
    const queryVariables = {};

    const TestRenderer = ({ componentProps, queryVariables: variables }) => {
      const data = useLazyLoadQuery<DashboardTitleTestQuery>(query, variables);
      if (!data) return null;
      return <DashboardTitle viewer={data.viewer} {...componentProps} />;
    };

    TestDashboardTitle = componentProps => (
      <RelaySuspensFragmentTest environment={environment}>
        <TestRenderer componentProps={componentProps} queryVariables={queryVariables} />
      </RelaySuspensFragmentTest>
    );

    environment.mock.queueOperationResolver(operation =>
      MockPayloadGenerator.generate(operation, defaultMockResolvers),
    );
  });

  describe('<TestDashboardTitle />', () => {
    it('should render correctly', () => {
      testComponentTree = ReactTestRenderer.create(<TestDashboardTitle />);
      expect(testComponentTree).toMatchSnapshot();
    });
  });
});
