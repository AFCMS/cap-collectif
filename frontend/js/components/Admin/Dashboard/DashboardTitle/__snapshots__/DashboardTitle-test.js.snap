// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DashboardTitle /> <TestDashboardTitle /> should render correctly 1`] = `
<div
  className="sc-bdnylx Flex___StyledAppBox-ll88x1-0 ixbuAQ"
>
  <p
    className="sc-bdnylx brZeUD"
  >
    dashboard-platform
  </p>
  <div
    className="sc-bdnylx Flex___StyledAppBox-ll88x1-0 bPbIam gqZsEv"
  >
    <p
      className="sc-bdnylx fNJNWw"
    >
      <span
        className="sc-bdnylx iRaWyl"
      >
        10
      </span>
      global.project-dynamic
    </p>
    <p
      className="sc-bdnylx fNJNWw"
    >
      <span
        className="sc-bdnylx jA-dhDL"
      >
        5
      </span>
      global.in-progress
    </p>
    <p
      className="sc-bdnylx fNJNWw"
    >
      <span
        className="sc-bdnylx dtwjrQ"
      >
        5
      </span>
      global.done-dynamic
    </p>
  </div>
</div>
`;
